source/common/grpc/buffered_async_client.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include "source/common/grpc/typed_async_client.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/container/btree_map.h"
+
+namespace Envoy {
+namespace Grpc {
+
+enum class BufferState { Buffered, PendingFlush };
+
+// This class wraps bidirectional gRPC and provides message arrival guarantee.
+// It stores messages to be sent or in the process of being sent in a buffer,
+// and can track the status of the message based on the ID assigned to each message.
+// If a message fails to be sent, it can be re-buffered to guarantee its arrival.
+template <class RequestType, class ResponseType> class BufferedAsyncClient {
+public:
+  BufferedAsyncClient(uint32_t max_buffer_bytes, const Protobuf::MethodDescriptor& service_method,
+                      Grpc::AsyncStreamCallbacks<ResponseType>& callbacks,
+                      const Grpc::AsyncClient<RequestType, ResponseType>& client)
+      : max_buffer_bytes_(max_buffer_bytes), service_method_(service_method), callbacks_(callbacks),
+        client_(client) {}
+
+  ~BufferedAsyncClient() {
+    if (active_stream_ != nullptr) {
+      active_stream_ = nullptr;
+    }
+  }
+
+  void bufferMessage(RequestType& message) {
+    const auto buffer_size = message.ByteSizeLong();
+    if (current_buffer_bytes_ + buffer_size > max_buffer_bytes_) {
+      return;
```

Comment:
In the logging case, I assume the logic is that we'll just drop the message if we hit the buffer limit. But for general use, should this return some indication that the message wasn't buffered and won't be sent?